### PR TITLE
test(language-service): add new mock host for testing ivy

### PR DESCRIPTION
**Pull request by kyliau**
This commit adds a new mock host for testing the ivy language service.

Unlike the existing mock_host which mocks the LanguageServiceHost, the
Ivy mock host mocks just the filesystem interface, aka ts.ServerHost.

This is because Ivy language service requires an actual Project to
perform operations like adding synthetic typecheck files to the project,
and by extension, to the ts.Program. These requirements make the existing
mock host unsuitable to be reused.

This new testing structure also improves test performance, because the
old mock host copies (it actually creates symlinks, but still that's
relatively expensive due to the sheer number of files involved) all
@angular/* packages along with the typescript package to a temporary
node_modules directory every time setup() is called.
Instead, this new mock host just loads them from a pre-determined path
in Bazel runfiles.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/language-service/test/project/tsconfig.json [hunk 1]
```diff
@@ -1,12 +1,18 @@
 {
-  "//00": "This file is used for IDE only, actual compilation options is in MockTypescriptHost in test_utils.ts",
+  "//00": "This file is used for both IDE and the actual Project creation in Ivy language service for testing purpose",
   "compilerOptions": {
-    "lib": ["es2015"],
+    "lib": ["es2015", "dom"],
     "strict": true,
     "experimentalDecorators": true,
+    "moduleResolution": "node",
```
Comment (ayazhafiz) [inline packages/language-service/test/project/tsconfig.json:7]: Why is this needed?

**Review by ayazhafiz — APPROVED** [review]

File: packages/language-service/ivy/test/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+)
+
+jasmine_node_test(
+    name = "test",
+    data = [
+        # Note that we used to depend on the npm_package of common, core, and
+        # forms, but this is no longer the case. We did it for View Engine
+        # because we wanted to load the flat dts, which is only available in the
+        # npm_package. Ivy does not currently produce flat dts, so we might
+        # as well just depend on the outputs of ng_module.
```
[the remark this answers is not in the document]
Comment (kyliau) [reply packages/language-service/ivy/test/BUILD.bazel:19]: the outputs are d.ts files with Ivy annotations.
It's basically what you get when you run `bazel build --config=ivy //packages/core`

File: packages/language-service/ivy/test/mock_host.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  readFile(absPath: string, encoding?: string): string |
+      undefined {
+        // TODO: Need to remove all annotations in templates like we do in
+        // MockTypescriptHost
+        return ts.sys.readFile(absPath, encoding);
+      },
+  // TODO: Need to cast as never because this is not a proper ServerHost interface.
+  // ts.sys lacks methods like watchFile() and watchDirectory(), but these are not
+  // needed for now.
+} as never;
```
[the remark this answers is not in the document]
Comment (kyliau) [reply packages/language-service/ivy/test/mock_host.ts:50]: They're effectively the same. Instead of doing what you suggested I set the type on line 39 to avoid casting twice.

File: packages/language-service/test/project/tsconfig.json (hunk 1 above)
Comment (kyliau) [reply packages/language-service/test/project/tsconfig.json:7]: Because we are now targeting "es2015", the default moduleResolution is "classic". [1].
But we want to include package.json in the resolution process, so we have to be explicit here.
Another reason is that we want this config to be as close as the default generated by Angular CLI.

[1]: https://www.typescriptlang.org/docs/handbook/compiler-options.html

File: packages/language-service/ivy/test/mock_host_spec.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      if (fileName.startsWith(TEST_SRCDIR)) {
+        return fileName.substring(TEST_SRCDIR.length);
+      }
+      return fileName;
+    });
+    expect(sourceFiles).toEqual(jasmine.arrayContaining([
+      // This shows that module resolution works
+      '/angular/packages/common/src/common.d.ts',
+      '/angular/packages/core/src/core.d.ts',
+      '/angular/packages/forms/src/forms.d.ts',
```
[the remark this answers is not in the document]
Comment (kyliau) [reply packages/language-service/ivy/test/mock_host_spec.ts:30]: > does this mean that the tests run in the /angular directory in the virtual file system created by Bazel?

Yes, that is how runfiles in Bazel works. (`/angular` is just the root of the Bazel workspace)

However, it is important to know that module resolution works even without Bazel in the mix.
The `baseUrl` combined with `paths` mapping basically tells the module resolver that
"whenever you see module names that match `@angular/*`, go look in `../../../../packages`".
For the IDE, this instructs the module resolver to go look at the actual sources.
For the test, this instructs the module resolver to go look at the d.ts files produced by ng_module.
The d.ts files are laid out by Bazel in a way that mimics the actual source structure.

[the remark this answers is not in the document]
Comment (ayazhafiz) [reply packages/language-service/ivy/test/mock_host_spec.ts:30]: Yeah, that makes sense. Is it safe to assume the runfiles directory will _always_ be `/angular`?

[the remark this answers is not in the document]
Comment (kyliau) [reply packages/language-service/ivy/test/mock_host_spec.ts:30]: yes, the name comes from the Bazel workspace
https://github.com/angular/angular/blob/12fcc7cafe8bf6dd33416847327891b581281774/WORKSPACE#L1-L4
it's very unlikely to change.

**Review by filipesilva — APPROVED** [review]
